components/support/rs_client/src/http_client.rs

```diff
@@ -37,53 +95,111 @@ pub(crate) enum RemoteState {
 
 impl Client {
     /// Create a new [Client] with properties matching config.
-    pub fn new(config: ClientConfig) -> Result<Self> {
+    /// During construction, the [Client] will fetch the base attachments
+    /// URL from the server's metadata, assuming attachments are enabled for
+    /// the server.
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
         let server_url = config
             .server_url
             .unwrap_or_else(|| String::from("https://firefox.settings.services.mozilla.com"));
         let bucket_name = config.bucket_name.unwrap_or_else(|| String::from("main"));
+        let base_url = Url::parse(&server_url)?;
+        let lookup = Self::lookup_attachments_base_url(base_url.clone());
```

Comment:
I definitely hear you on sidestepping the complexity of mutexes and optionals to do this the first time an attachment is downloaded!

There are a couple of reasons why I'm nervous about looking up the base URL at client creation time, though:

* `Client::new` now hides a blocking network call that can take an unknown amount of time to complete. In general, I try to avoid constructors doing I/O, or having side effects beyond throwing / returning errors for invalid parameters. Effectful constructors can make testing harder, and require consumers to remember to call the constructor off the main thread. The complexity is still there, but it's pushed to consumers.
* The network request can take an unknown amount of time, blocking whatever thread creates the `Client` until it returns. Consumers will need to remember to call the constructor off the main thread
* What happens if the request fails—maybe the server had a hiccup, or the user is on a spotty network? If we look up the base URL lazily, when an attachment is downloaded, we can bubble that error up, and try again on the next `get_attachment` call.
* Every consumer of `Client` has to wait for the network call, even if they never call `get_attachment`.